### PR TITLE
Add optional page/line/word bounding boxes for PDF & image inputs (--emit-bbox) + evaluation notes

### PR DESCRIPTION
**Pull request by mapo80**
# Add optional page/line/word bounding boxes for PDF & image inputs (`--emit-bbox`) + evaluation notes

## Summary

This PR adds an **opt-in** capability to emit page-anchored **bounding boxes** for text extracted from **PDFs and images**. When enabled, MarkItDown produces a sidecar JSON file that contains:

* page geometry
* line-level boxes and their span offsets into the generated Markdown
* word-level boxes with optional OCR confidence values

The feature is off by default and **does not change any existing outputs**. It is designed for downstream tasks that need spatial grounding—e.g., overlaying selections, table cell alignment checks, redaction previews, or training data generation for doc-layout models.

The fork also documents **performance measurements on the Docling test set** and a **delta analysis vs Docling’s outputs**, to help frame accuracy/robustness and cost. ([[GitHub](https://github.com/mapo80/markitdown)][1])

---

### New CLI flags

* `--emit-bbox`
  When present, MarkItDown writes a sidecar JSON file next to the Markdown output (e.g., `sample.pdf` → `sample.bbox.json`). Applies to **PDF** and **image** inputs. For image-only or scanned PDFs (no text layer), bounding boxes are obtained via OCR. ([[GitHub](https://github.com/mapo80/markitdown)][1])

* `--ocr-lang <lang-codes>` (optional)
  Controls OCR language(s) for cases where OCR is used. Mirrors `MARKITDOWN_OCR_LANG` (see Env Vars). ([[GitHub](https://github.com/mapo80/markitdown)][1])

### Environment variables

* `MARKITDOWN_OCR_LANG` – default OCR language(s) when `--emit-bbox` triggers OCR.
* `TESSDATA_PREFIX` – path to custom tessdata if needed.
  (Both only matter when OCR is used, i.e., scanned PDFs / images.) ([[GitHub](https://github.com/mapo80/markitdown)][1])

> Upstream currently **does not expose per-line / per-token coordinates**, which is a common user ask. This PR proposes a minimal, backwards-compatible way to add that capability. ([[GitHub](https://github.com/microsoft/markitdown/issues/1324?utm_source=chatgpt.com)][2])

---

## Output format (sidecar JSON)

When `--emit-bbox` is set, MarkItDown writes `<basename>.bbox.json` with the following structure:

```json
{
  "version": "1.0",
  "source": "sample.pdf",
  "pages": [
    { "page": 1, "width": 612, "height": 792 }
  ],
  "lines": [
    {
      "page": 1,
      "text": "Hello",
      "bbox_norm": [0, 0, 0, 0],
      "bbox_abs":  [0, 0, 0, 0],
      "confidence": null,
      "md_span": { "start": null, "end": null }
    }
  ],
  "words": [
    {
      "page": 1,
      "text": "Hello",
      "bbox_norm": [0, 0, 0, 0],
      "bbox_abs":  [0, 0, 0, 0],
      "confidence": null,
      "line_id": 0
    }
  ]
}
```

**Semantics**

* `bbox_abs` is in **pixel units of the page/image**, **top-left origin**.
* `bbox_norm` is **normalized** `[x0, y0, x1, y1]` in `[0,1]` relative to page width/height.
* `md_span` links a **line** back into `result.text_content` via **character offsets** (`start`, `end`), enabling exact highlighting in the Markdown string.
* `line_id` associates **word** items with their parent line.
* `confidence` is `null` when unavailable (e.g., embedded text), or a numeric value when the OCR engine returns one. ([[GitHub](https://github.com/mapo80/markitdown)][1])

---

## How it works (high level)

* **PDF with text layer**: Extract the text and geometry from the PDF’s text layer, build line- and word-level boxes, and record their corresponding spans in the produced Markdown.
* **Scanned PDFs / Images**: When no text layer is present, run OCR (Tesseract) for text + geometry; languages can be set via `--ocr-lang` / `MARKITDOWN_OCR_LANG`. `TESSDATA_PREFIX` is respected for custom language packs. ([[GitHub](https://github.com/mapo80/markitdown)][1])

This design keeps the default UX unchanged and only introduces extra work when explicitly requested.

---

## Performance notes (Docling test data)

To help reviewers understand the runtime impact, the fork includes a small **timing study** on the **Docling test dataset** (12 documents across PDF/PNG/TIFF). Highlights:

* **Average** Markdown conversion time: **3.18 s**
* **Average** bbox generation time (with `--emit-bbox`): **5.10 s**
* By type (avg MD / avg BBox):

  * **PDF**: 3.29 s / 5.14 s
  * **PNG**: 2.51 s / 5.56 s
  * **TIFF**: 2.57 s / 4.19 s

Full per-file table is included in the README section of the fork. ([[GitHub](https://github.com/mapo80/markitdown)][1])

---

## Accuracy / quality observations

On the same Docling test set, a simple delta analysis compares MarkItDown outputs to Docling’s ground truth:

* **Markdown content delta**: \~**45%** average difference
* **Bounding-box coordinate deviation**: \~**18%** average

Bigger discrepancies were seen on **right-to-left pages** and **scanned forms**; these are called out for future iteration. (Context and numbers are documented in the fork’s README.) ([[GitHub](https://github.com/mapo80/markitdown)][1])

The fork also includes comparison notes against FUNSD (see `funsd_bbox_comparison.md`) to illustrate layout alignment behavior on form-like documents. ([[GitHub](https://github.com/mapo80/markitdown)][1])

---

## Why this belongs in MarkItDown

* **Spatial grounding** is increasingly required for RAG, redaction, annotation tools, and evaluation pipelines.
* The feature is **opt-in** and unobtrusive; it does not alter existing Markdown outputs or require new mandatory dependencies.
* It fills a well-documented gap—upstream does not currently ship span/box geometry. ([[GitHub](https://github.com/microsoft/markitdown/issues/1324?utm_source=chatgpt.com)][2])

---

## Backwards compatibility

* Default behavior unchanged.
* Sidecar JSON is emitted **only** when `--emit-bbox` is provided.
* OCR is invoked only when needed (images or PDFs lacking a text layer). ([[GitHub](https://github.com/mapo80/markitdown)][1])

---

## Documentation added

The fork’s README adds a **“Bounding Boxes”** section with:

* the CLI usage (`--emit-bbox`)
* schema explanation (absolute vs normalized boxes, top-left origin, `md_span`)
* OCR language/env configuration
* links to the Docling comparison docs and timing table

These doc bits can be ported verbatim or adapted into the upstream README. ([[GitHub](https://github.com/mapo80/markitdown)][1])

---

## Testing & metrics (what reviewers can look at)

* **Schema sanity checks** (implicit in the docs):

  * coordinates clamped to page, `x0 ≤ x1`, `y0 ≤ y1`;
  * normalized boxes in `[0,1]`;
  * `md_span` ranges map back to the same text content as the corresponding line.
* **Benchmarks included** in README: per-file timings + averages; delta analysis vs Docling ground truth (content diff and coord deviation). ([[GitHub](https://github.com/mapo80/markitdown)][1])

> References
> • Fork README “Bounding Boxes”, Docling comparisons, and timing table. ([[GitHub](https://github.com/mapo80/markitdown)][1])
> • Docling project (test dataset used for comparisons). ([[GitHub](https://github.com/docling-project/docling?utm_source=chatgpt.com)][3])
> • Upstream issue commentary noting lack of coordinate exposure. ([[GitHub](https://github.com/microsoft/markitdown/issues/1324?utm_source=chatgpt.com)][2])

---

## Limitations & next steps (proposed follow-ups)

* Improve robustness on **RTL** and **scanned forms** (noted as the main sources of deviation). ([[GitHub](https://github.com/mapo80/markitdown)][1])
* Consider emitting **token-to-Markdown** mapping for **words** (currently `md_span` is provided at line-level), enabling finer-grained highlight sync.
* Optional **polygon** support for non-axis-aligned text (e.g., rotated scans).
* Expand test corpus beyond Docling to include more invoices/forms (e.g., FUNSD) with IoU-based metrics in CI.

---

## Checklist

* [x] Feature is **opt-in** (`--emit-bbox`), default behavior unchanged
* [x] Sidecar schema documented (abs/normalized coords, md spans, confidence)
* [x] OCR language and tessdata path can be configured via flag/env vars
* [x] Performance and accuracy notes included (Docling test set)
* [x] No breaking API changes

---

### Screenshots / Examples

Run:

```bash
markitdown sample.pdf --emit-bbox
# Produces: sample.md  and  sample.bbox.json
```

Excerpt from the emitted JSON is shown above; see the fork README for full details and per-file timings. ([[GitHub](https://github.com/mapo80/markitdown)][1])

---

**Thanks for reviewing!** Happy to split this into a docs-only PR + feature PR if you prefer, or to iterate on schema details (e.g., add per-word `md_span`, include page DPI, or expose rotation angles).

[1]: https://github.com/mapo80/markitdown "GitHub - mapo80/markitdown: Python tool for converting files and office documents to Markdown."
[2]: https://github.com/microsoft/markitdown/issues/1324?utm_source=chatgpt.com "Does it support the text recognization? · Issue #1324"
[3]: https://github.com/docling-project/docling?utm_source=chatgpt.com "docling-project/docling: Get your documents ready for gen AI"

### CODE REVIEW

**Comment by mapo80** [comment]
@microsoft-github-policy-service agree

**Comment by mapo80** [comment]
> @mapo80 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree
